### PR TITLE
Makes UpsertSecurityGroup stage/task generic.

### PR DESCRIPTION
**Pull request by ttomsu**
This one's a doosey...

Security group creation/updating is complex, so I took the same approach I took with the server group creation. Namely, I use a single generic stage + task, and delegate out to specific providers for the detailed creation of the operations. 

Each provider also used a specific `WaitForUpsertedSecurityGroupTask`, so I used the same provider-specific `SecurityGroupUpserter` object to encapsulate the logic to verify that the security group has been updated. 

Testing was much more straightforward than I expected, because I was able to save the bulk of the tests (which are also complex!) with little modification. 

Tested against both AWS and GCE, as well as the Clone Security Group operation, which reuses `WaitForUpsertedSecurityGroupTask`. 

@cfieber @ajordens @duftler PTAL

@spinnaker/reviewers FYI

### CODE REVIEW

**Comment by ttomsu** [comment]
Ping

**Comment by duftler** [comment]
LGTM.
@ajordens or @cfieber any objection to us merging this?

**Comment by duftler** [comment]
Rebasing now...

**Comment by ajordens** [comment]
Just taking a look

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/SecurityGroupUpserter.groovy [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+  /**
+   * @return a two item list.
+   *
+   * First item is a list of operation descriptors. Each operation should be a single entry map keyed by the operation name,
+   * with the operation map as the value. (A list of maps of maps? We must go deeper...)
+   *
+   * Second item is a Map<String, Object> to be inserted into the stage's output map.
+   */
+  OperationContext getOperationContext(Stage stage)
```
Comment (ajordens) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/SecurityGroupUpserter.groovy]: Comment above doesn't seem to jive with `OperationContext`.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/SecurityGroupUpserter.groovy [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  boolean isSecurityGroupUpserted(MortService.SecurityGroup upsertedSecurityGroup, Stage stage)
+
+  /**
+   * @return The cloud provider type that this object supports.
+   */
+  String getCloudProvider()
+
+  @Canonical
+  static class OperationContext {
+    List<Map> opertations
```
Comment (ajordens) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/SecurityGroupUpserter.groovy]: `opertations` vs. `operations`

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/SecurityGroupUpserter.groovy (hunk 1 above)
Comment (ttomsu) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/SecurityGroupUpserter.groovy]: Nice catch - fixed.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/SecurityGroupUpserter.groovy (hunk 2 above)
Comment (ttomsu) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/SecurityGroupUpserter.groovy]: D'oh! Fixed.

**Comment by ajordens** [comment]
LGTM

**Pull request merged**